### DISCUSSION TITLE
Multi-collection in a RAG system

### DISCUSSION THREAD

**Discussion by wilsoncastiblanco** [milvus-io/milvus#32876] [Q&A and General discussion] [answered]
Hey there!!

I'm building a RAG system using a different kind of documents, initially, I created a collection with a knowledge base of JSON files, but then I tried to insert into the same collection individual PDF files, so I got this error trying to store the chunks into the vector db using `from_documents` 

Error: `pymilvus.exceptions.DataNotMatchException: <DataNotMatchException: (code=1, message=The data don't match with schema fields, expect 12 list, got 3)>`

This error surely is because the collection was created with an initial schema based on the JSON files (?) and when I try to add the PDF chunks it fails.

Reading the documentation I found that I can create different collections, but then, how can I search on the main JSON collection and at the same time in the PDF collection to get an accurate response in my RAG system?

Is there a way to accomplish that? I appreciate your thoughts on this.

**Comment by yhmo** [comment, marked as answer]
Looks like you are using langchain.
The from_documents() is an interface of langchain.VectorStore. It accepts a list of langchian.Document and an embeddings callback function:
```
def from_documents(
        cls: Type[VST],
        documents: List[Document],
        embedding: Embeddings,
        **kwargs: Any,
    )
```
Each langchain.Document has two members: page_content and metadata. page_content is a text content of your document, metadata is a dict which contains the extra properties of the document.
```
class Document(Serializable):
    page_content: str
    metadata: dict = Field(default_factory=dict)
```

When you first time call the from_documents(), it initializes a Milvus middleware to create a new collection in Milvus:
https://github.com/langchain-ai/langchain/blob/9992beaff9205825993ca65f589e9661bcadd939/libs/community/langchain_community/vectorstores/milvus.py#L294

By default, it deduces a collection schema according to the first metadata you input:
https://github.com/langchain-ai/langchain/blob/9992beaff9205825993ca65f589e9661bcadd939/libs/community/langchain_community/vectorstores/milvus.py#L323

If the second metadata is different to the first, you will get the error "=The data don't match with schema fields"

But if you specify a parameter "metadata_field", it will define a JSON field to store all the metadata in JSON format:
https://github.com/langchain-ai/langchain/blob/9992beaff9205825993ca65f589e9661bcadd939/libs/community/langchain_community/vectorstores/milvus.py#L131

With the JSON metadata field, you can input different metadata when you call the from_documents()
`from_documents(documents = xxxx, embeddings=xxxx, metadata_field="META_FIELD")`

**Reply by wilsoncastiblanco** [reply]
Hey @yhmo thanks for answering. Let me try that. I'll let you know how it goes.

**Reply by wilsoncastiblanco** [reply]
@yhmo I tested it out last night and it worked!! 🎉 but, I'm using `partition_key_field` so by using `metadata_field` there was another error saying that the `partition_key_field` should be VARCHAR or INT64, and it is because internally if the `metadata_field` is provided, then the metadata is converted into JSON 
```
if self._metadata_field is not None:
            fields.append(FieldSchema(self._metadata_field, DataType.JSON))
        else:
```
so based on your suggestion, I took only some important fields to match the metadata and it worked that way because it is the same metadata for all the documents

```
        doc.metadata = {
            "source": doc.metadata["source"], 
            "file_path": doc.metadata["file_path"], 
            "page": doc.metadata["page"], 
            "namespace": id
            }
```
I removed the `metadata_field` having standardized the metadata. That was the initial problem, the document metadata was not matching, and some of them had fewer or more fields.

Thank you so much for your help!

**Reply by yhmo** [reply]
Correct, if you are using partition_key_field, the metadata must be standardized. It deduces the schema by the first value type of each meta.

**Reply by JenuelDev** [reply]
How to do query for `Partition-key-based`. Lets say for example I have a partition key field called 'tenant' and it contains INT type like the ID of the user from mysql database. How to query ?

```python
collection = MilvusClient()

collection.search(
 'collection_name' => 'myCollection',
 'tenant' => 46 # the user id
)
```
I know the code will have an error above, but I think it shows what I am trying to search. Thanks thanks :)

**Reply by yhmo** [reply]
Internal machinery of partition-key:
1. the collection has a partition-key, INT type, let's say the field name is "tenant". There is a parameter num_partitions to defines the number of partitions in this collection, default value is 64. Once the collection is created, the number of partitions cannot be changed.
2. user calls insert() to insert 1000 rows into the collection, milvus calculates the hash values of each "tenant", and stores the rows into different partitions according to the mod(hash_vaue, num_partitions)
3. partition-key helps improving search performance only when your filtering expression contains the partition-key. If you search with collection.search(expr="tenant == 8"), the search performance is best because it calculates the hash value of 8 and quickly knows which partition the tenant=8 is stored, then only search one partition to get the result. But if you search with collection.search(expr="id ==9"), the partition-key won't help performance.

**Reply by wilsoncastiblanco** [reply]
@yhmo `id` is always assigned to a document? 

I'm not sure if this is the expected answer for @JenuelDev, but Milvus contains expressions to get the data by the partition key defined https://milvus.io/docs/v2.2.x/partition_key.md#Conduct-an-ANN-Search, which is the 3rd point explained by @yhmo.

btw, @yhmo is the expert here :D

**Reply by JenuelDev** [reply]
@yhmo can you check this.. so I am using the search function using the milvus Client will this work?
```python
def search_and_query(
        self,
        search_vectors: list,
        search_field: str = "embeddings",
        search_params={"metric_type": "L2", "params": {"nprobe": 10}},
        vector_search_limit=10,
    ):
        self.milvusClient.load_collection(self.collection_name)
        
        params = {
            "collection_name":self.collection_name,
            "data":search_vectors,
            "search_params":search_params,
            "anns_field":search_field,
            "limit":vector_search_limit,
            "output_fields":["source"],
            "consistency_level":"Strong",
        }
        
        if self.tenant is not None:
            params["filter"] = f"tenant=={self.tenant}"
        
        result = self.milvusClient.search(**params)

        return result
```

**Reply by yhmo** [reply]
params["filter"] = f"tenant=={self.tenant}" --------------------- Yes, it works. Since the params contains "filter" and the filtering expression contains the partition key.

**Reply by yhmo** [reply]
@wilsoncastiblanco 
I just give the example of  "filtering expression contains the partition-key" and "filtering expression doesn't contain the partition-key"
- Assume "id" is a field of the collection but is not a partition key field. "id ==9" doesn't contain the partition key. The Search will scan all data in the collection to find the item whose "id" is 9.
- "tenant" is a partition key field.   "tenant == 8" contains the partition key. The Search will scan only one partition in which the tenant=8 is.

**Comment by yhmo** [comment]
For a RAG system, no need to create lots of collections. Typically, one collection is ok.
You input documents batch by batch, the VectorStore calls the embeddings callback function to generate vectors from the page_content. Then store the metadata into metadata_field. When you call similarity_search() interface, it will call the search() interface of Milvus and return the topk results along with their metadata. You even no need to care what format the metadata is stored inside the Milvus.

**Reply by wilsoncastiblanco** [reply]
After understanding how that works after your explaining, I only have one collection and it worked.

**Reply by JenuelDev** [reply]
@yhmo is metadata in milvus automatic? or should be set when setting the schemma?

**Reply by yhmo** [reply]
@JenuelDev 
Milvus supports these scalar field types:
- BOOL: Boolean (true or false)
- INT8: int8
- INT16: int16
- INT32: int32
- INT64: int64
- FLOAT: float32
- DOUBLE: double
- VARCHAR: VARCHAR
- [JSON](https://milvus.io/docs/use-json-fields.md): JSON
- [Array](https://milvus.io/docs/array_data_type.md): Array

You can define multiple fields to store metadata. It it recommended to store a metadata in an independent field if this metadata will take part in the search. For example, I have a metadata named "age", I will use an INT8 field to store "age" values, so that I can search with filtering expressions such as "age > 20".
If you don't need to filter on a metadata, you can enable the collection's dynamic schema so that all the undefined metadata is stored in dynamic field.
Dynamic field is a special/hidden JSON field in the collection, all the undefined metadatd is stored in it in JSON format. Filtering performance is low on JSON field.
https://milvus.io/docs/enable-dynamic-field.md
